### PR TITLE
Feat/support moe

### PR DESCRIPTION
**Pull request by kilinchange**
引入 Megatron 风格的 MoE 基础训练路径，并提供 tiny_mixtral 端到端验证样例。打通单卡 MoE 训练流程，为后续 AllToAll EP、Grouped GEMM expert、真实 Mixtral 权重转换做基础铺垫。

Megatron Moe 模块调研：https://gxtctab8no8.feishu.cn/docx/Ry6XdG8Dro0PoSxc7NscNa4WnFf?from=from_copylink

### 主要修改

1. 新增 MoE 结构相关模块：

- TopKRouter 模块：logits -> score_function -> topk -> scatter，输出 dense routing_probs 和 bool routing_map，支持 top-k routing、top-k 权重归一化、scaling factor 等配置
- Experts 模块：专家计算层
   - SequentialMLP：最直接的基线实现，逐个 expert 顺序执行
- MoETokenDispatcher 模块
  - MoEAllGatherTokenDispatcher：单卡退化路径
- MoELayer 模块：组装 TopKRouter、Token Dispatcher、experts 三个模块，替换原有 dense 网络结构的 MLP 层计算

2. 新增 tiny_mixtral 示例：

- 文件结构对齐 example/llama3
- 支持从 LLMC .bin 加载权重
- 使用真实 token 数据训练
- config 尽量贴近 Mixtral 8x7B，未对齐字段保留简短注释说明

3. 测试和验证：

- 增加 MoE layer/top-k 相关架构测试
- 集成 tiny_mixtral 到测试/profile 流程
- 增加 Megatron tiny Mixtral 对齐脚本（待合入测试仓库）和日志留档，用于确认 loss 对齐

### CODE REVIEW

File: scripts/test_config.json
Comment (kilinchange) [inline scripts/test_config.json:26]: moe 放 test_groups 下的 tag，.sh 里做筛选。

Comment (kilinchange) [reply scripts/test_config.json:26]: done

File: example/tiny_mixtral/main.cc [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        dtype = infini_train::DataType::kFLOAT32;
+    } else if (FLAGS_dtype == kDtypeBF16) {
+        dtype = infini_train::DataType::kBFLOAT16;
+    } else {
+        LOG(FATAL) << "Datatype " << FLAGS_dtype << " not supported.";
+    }
+
+    auto loss_fn = std::make_shared<infini_train::nn::CrossEntropyLoss>();
+    auto optimizer
+        = infini_train::optimizers::SGD::Create(static_cast<float>(FLAGS_learning_rate))(model->Parameters());
```
Comment (kilinchange) [inline example/tiny_mixtral/main.cc]: 用 adam 优化器。

File: infini_train/src/kernels/cuda/topk.cu [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            top_values[out_offset] = input[outer * dim_size * inner_size + idx * inner_size + inner];
+            top_indices[out_offset] = idx;
+        }
+    }
+}
+
+std::vector<std::shared_ptr<Tensor>> TopKForward(const std::shared_ptr<Tensor> &input, int64_t topk, int64_t dim,
+                                                 bool largest, bool sorted) {
+    CHECK_GE(input->Dims().size(), 1);
+    (void)sorted;
```
Comment (kilinchange) [inline infini_train/src/kernels/cuda/topk.cu]:
```
CHECK(sorted)
    << "TopK currently only supports sorted=true";
```

File: infini_train/src/kernels/cpu/topk.cc [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+#include "infini_train/include/dispatcher.h"
+#include "infini_train/include/tensor.h"
+
+namespace infini_train::kernels::cpu {
+
+std::vector<std::shared_ptr<Tensor>> TopKForward(const std::shared_ptr<Tensor> &input, int64_t topk, int64_t dim,
+                                                 bool largest, bool sorted) {
+    CHECK(input->Dtype() == DataType::kFLOAT32) << "CPU TopKForward currently supports float32 only";
+    CHECK_GE(input->Dims().size(), 1);
+    (void)sorted;
```
Comment (kilinchange) [inline infini_train/src/kernels/cpu/topk.cc]:
```
CHECK(sorted)
    << "TopK currently only supports sorted=true";
```

File: infini_train/include/nn/modules/transformer/moe/moe_utils.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "infini_train/include/nn/modules/transformer/transformer_config.h"
+#include "infini_train/include/tensor.h"
+
+namespace infini_train::nn::moe {
+
+struct PermutationMetadata {
```
Comment (kilinchange) [inline infini_train/include/nn/modules/transformer/moe/moe_utils.h:17]: 补注释，对齐 megatron 命名

File: infini_train/src/nn/modules/transformer/mlp.cc [hunk 5]
```diff
@@ -37,6 +37,12 @@ MLP::MLP(const TransformerConfig &config) : CloneableModule(kType) {
     // Round up to multiple_of
     ffn_hidden = (ffn_hidden + config.multiple_of - 1) / config.multiple_of * config.multiple_of;
 
+    if (config.ffn_type == FFNType::kMoE && config.moe_config.has_value()
+        && config.moe_config->moe_ffn_hidden_size > 0) {
+        ffn_hidden = config.moe_config->moe_ffn_hidden_size;
```
Comment (kilinchange) [inline infini_train/src/nn/modules/transformer/mlp.cc]: 禁止静默跳过，
```
if (config.ffn_type == FFNType::kMoE) {
    const auto& moe_config = RequireMoEConfig(config);
    CHECK_GT(config.moe_config->moe_ffn_hidden_size, 0);
    ffn_hidden = config.moe_config->moe_ffn_hidden_size;
}
```

File: infini_train/src/kernels/cpu/topk.cc [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    const auto *idx_ptr = static_cast<const int64_t *>(indices->DataPtr());
+    for (int64_t outer = 0; outer < outer_size; ++outer) {
+        for (int64_t inner = 0; inner < inner_size; ++inner) {
+            for (int64_t selected = 0; selected < topk; ++selected) {
+                const int64_t out_offset = outer * topk * inner_size + selected * inner_size + inner;
+                const int64_t selected_idx = idx_ptr[out_offset];
+                CHECK_GE(selected_idx, 0);
+                CHECK_LT(selected_idx, dim_size);
+                std::memcpy(dst + (outer * dim_size * inner_size + selected_idx * inner_size + inner) * elem_size,
+                            src + out_offset * elem_size, elem_size);
```
Comment (JYMiracle305) [inline infini_train/src/kernels/cpu/topk.cc]: 这里建议命名成 src_offset，另外dst的偏移也提成 dst_offset，读起来好理解

File: infini_train/include/nn/modules/transformer/transformer_config.h [hunk 7]
```diff
@@ -15,11 +15,45 @@ enum class MLPType {
...
+
+    int64_t num_experts = 0;
+    int64_t expert_parallel_size = 1;
+    int64_t router_topk = 1;
+    bool router_pre_softmax = false;
+    std::optional<float> router_topk_scaling_factor = std::nullopt;
+    RouterScoreFunction router_score_function = RouterScoreFunction::kSoftmax;
+    float aux_loss_coeff = 0.0f;
+    std::optional<float> expert_capacity_factor = std::nullopt;
+    bool pad_expert_input_to_capacity = false;
```
Comment (JYMiracle305) [inline infini_train/include/nn/modules/transformer/transformer_config.h]: 这里有没用到的配置，建议在用到的时候再加

File: example/tiny_mixtral/main.cc [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+#include <iostream>
+#include <memory>
+#include <vector>
+
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+
+#include "example/common/tiny_shakespeare_dataset.h"
+#include "example/tiny_mixtral/checkpoint_loader.h"
+#include "example/tiny_mixtral/config.h"
```
Comment (JYMiracle305) [inline example/tiny_mixtral/main.cc]: example 下的头文件放置位置不对

File: infini_train/include/nn/modules/transformer/moe/moe_utils.h [hunk 9]
```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "infini_train/include/nn/modules/transformer/transformer_config.h"
+#include "infini_train/include/tensor.h"
```
Comment (JYMiracle305) [inline infini_train/include/nn/modules/transformer/moe/moe_utils.h]: 这里 Tensor 可以用前置声明

File: infini_train/include/nn/modules/transformer/moe/token_dispatcher.h [hunk 10]
```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "infini_train/include/nn/modules/transformer/moe/moe_utils.h"
+#include "infini_train/include/nn/modules/transformer/transformer_config.h"
+#include "infini_train/include/tensor.h"
```
Comment (JYMiracle305) [inline infini_train/include/nn/modules/transformer/moe/token_dispatcher.h]: 这个也可以前置声明

**Comment by kilinchange** [comment]
megatron 精度对比 pr：https://github.com/InfiniTensor/InfiniTrain-Test/pull/5
InfiniTrain 精度运行情况（node24 上运行得到）：
tiny_mixtral_1 (float32，小规模)
image
tiny_mixtral_1_bfloat16 (bfloat16，小规模)
image
tiny_mixtral_2 (float32，大规模)
image
tiny_mixtral_2_bfloat16 (bfloat16，大规模)
image

File: infini_train/src/kernels/cpu/topk.cc (hunk 6 above)
Comment (kilinchange) [reply infini_train/src/kernels/cpu/topk.cc]: done

File: infini_train/include/nn/modules/transformer/transformer_config.h (hunk 7 above)
Comment (kilinchange) [reply infini_train/include/nn/modules/transformer/transformer_config.h]: done

File: example/tiny_mixtral/main.cc (hunk 8 above)
Comment (kilinchange) [reply example/tiny_mixtral/main.cc]: done

File: infini_train/include/nn/modules/transformer/moe/moe_utils.h (hunk 9 above)
Comment (kilinchange) [reply infini_train/include/nn/modules/transformer/moe/moe_utils.h]: done

File: infini_train/include/nn/modules/transformer/moe/token_dispatcher.h (hunk 10 above)
Comment (kilinchange) [reply infini_train/include/nn/modules/transformer/moe/token_dispatcher.h]: done

**Comment by kilinchange** [comment]
g3025 测试结果
精度测试：
image
性能测试：
image

**Review by JYMiracle305 — APPROVED** [review]

**Comment by kilinchange** [comment]
合入前最终测试
精度测试结果：
image

性能测试结果：
image
